lib/ClangImporter/ImportType.cpp

```diff
@@ -1156,208 +1195,246 @@ static bool isNSString(Type type) {
   return false;
 }
 
-static ImportedType adjustTypeForConcreteImport(
-    ClangImporter::Implementation &impl,
-    ImportResult importResult, ImportTypeKind importKind,
-    bool allowNSUIntegerAsInt, Bridgeability bridging, OptionalTypeKind optKind,
-    bool resugarNSErrorPointer) {
-  Type importedType = importResult.AbstractType;
-  ImportHint hint = importResult.Hint;
-
-  if (importKind == ImportTypeKind::Abstract) {
-    return {importedType, false};
-  }
-
-  // 'void' can only be imported as a function result type.
-  if (hint == ImportHint::Void &&
-      (importKind == ImportTypeKind::AuditedResult ||
-       importKind == ImportTypeKind::Result)) {
-    return {impl.getNamedSwiftType(impl.getStdlibModule(), "Void"), false};
-  }
-
-  // Import NSString * globals as String.
-  if (hint == ImportHint::ObjCBridged && isNSString(importedType) &&
-      (importKind == ImportTypeKind::Variable ||
-       importKind == ImportTypeKind::AuditedVariable)) {
-    return {hint.BridgedType, false};
-  }
+static Type maybeImportNSErrorOutParameter(ClangImporter::Implementation &impl,
+                                           Type importedType,
+                                           bool resugarNSErrorPointer) {
+  PointerTypeKind PTK;
+  auto elementType = importedType->getAnyPointerElementType(PTK);
+  if (!elementType || PTK != PTK_AutoreleasingUnsafeMutablePointer)
+    return Type();
 
-  // For anything else, if we completely failed to import the type
-  // abstractly, give up now.
-  if (!importedType)
-    return {Type(), false};
+  auto elementObj = elementType->getOptionalObjectType();
+  if (!elementObj)
+    return Type();
 
-  // Special case AutoreleasingUnsafeMutablePointer<NSError?> parameters.
-  auto maybeImportNSErrorPointer = [&]() -> Type {
-    if (importKind != ImportTypeKind::Parameter)
-      return Type();
+  auto elementClass = elementObj->getClassOrBoundGenericClass();
+  if (!elementClass)
+    return Type();
 
-    PointerTypeKind PTK;
-    auto elementType = importedType->getAnyPointerElementType(PTK);
-    if (!elementType || PTK != PTK_AutoreleasingUnsafeMutablePointer)
-      return Type();
+  // FIXME: Avoid string comparison by caching this identifier.
+  if (elementClass->getName().str() !=
+        impl.SwiftContext.getSwiftName(KnownFoundationEntity::NSError))
```

Comment:
`getName()` returns an `Identifier` here, doesn't it? So can't you use `getSwiftId()` here instead of `getSwiftName()` to avoid stringifying?

Reply:
This code predates `getSwiftId`. Honestly not sure if doing the string compare every time is slower than the time to intern the string anyway, if it's not going to be cached locally per the (ancient) FIXME.